contracts/ServiceNodeRewards.sol

```diff
@@ -655,24 +658,45 @@ contract ServiceNodeRewards is Initializable, Ownable2StepUpgradeable, PausableU
     ///
     /// @return id The ID allocated for the service node and a pointer to its storage (equivalent to
     /// `_serviceNodes[id]`).
-    function serviceNodeAdd(BN256G1.G1Point memory pubkey) internal returns (uint64 id, ServiceNode storage sn) {
+    function serviceNodeAdd(BN256G1.G1Point memory blsPubkey, uint256 ed25519Pubkey) internal returns (uint64 id, ServiceNode storage sn) {
+
+
+        // NOTE: Check keys
+        if (blsPubkey.X == 0 || blsPubkey.Y == 0)
+            revert NullBLSPubkey();
+
+        if (ed25519Pubkey == 0)
+            revert NullEd25519Pubkey();
+
         // NOTE: Check if the service node already exists
         // (e.g. <BLS Key> -> <SN> mapping)
-        bytes memory pubkeyBytes = BN256G1.getKeyForG1Point(pubkey);
-        if (serviceNodeIDs[pubkeyBytes] != LIST_SENTINEL) revert BLSPubkeyAlreadyExists(serviceNodeIDs[pubkeyBytes]);
+        bytes memory blsPubkeyBytes = BN256G1.getKeyForG1Point(blsPubkey);
+
+
+        // NOTE: Check that keys aren't reused from any currently active Session
+        // nodes.
+        if (serviceNodeIDs[blsPubkeyBytes] != LIST_SENTINEL)
+            revert BLSPubkeyAlreadyExists(serviceNodeIDs[blsPubkeyBytes]);
+
+        if (blsToEd25519Pubkey[blsPubkeyBytes] != 0)
+            revert BLSPubkeyHasPairedEd25519Pubkey(blsToEd25519Pubkey[blsPubkeyBytes]);
+
+        BN256G1.G1Point storage prevBLSPubkey = ed25519ToBLSPubkey[ed25519Pubkey];
+        if (prevBLSPubkey.X != 0 || prevBLSPubkey.Y != 0)
+            revert Ed25519PubkeyHasPairedBLSPubkey(prevBLSPubkey);
```

Comment:
This logic feels a little bit weird because it's checking what I think is an impossible case where the bls pubkey doesn't exist, and yet there is an entry for the bls pubkey in the `bls->ed` mapping.

I think this implication could allow a couple of contract storage reduction optimizations by moving the ed25519 pubkey value into the ServiceNode struct:
- add `ServiceNode.ed25519Pubkey`; set it when adding a service node.
- drop the `blsToEd25519Pubkey` mapping
  - drop the `if (blsToEd25519Pubkey[blsPubkeyBytes] != 0)` check here.  (because not being in serviceNodeIDs already means there won't be a matching bls->ed pairing)
  - in `serviceNodeDelete` where we get the ed25519 pubkey out of the map for cleanup of the ed mapping, just get it from `node.ed25519Pubkey` instead.
- we can also reduce the storage of the ed -> bls pubkey mapping by mapping ed -> service node id instead.  Then in the last bit of the code I'm commenting on here (which I think is the only place we actually use the value) we can just check that this value against 0: if non-zero, we fail, and get the `prevBLSPubkey` out of `serviceNodeIDs` with the non-zero value.

Reply:
All valid points.

- Removed the `bls->ed` mapping as it is redundant
- Add `ed25519Pubkey` to `ServiceNode` struct to couple them together
- Renamed `ServiceNode.pubkey` to `ServiceNode.blsPubkey` to be more explicit now that we store 2 keys in the struct'

Simplifies the whole dupe checking business. 

One side effect of this is that it should make upgrading the current stagenet contract impossible because we didn't reserve any bytes in `ServiceNode` struct for it to update in-place. I did alleviate it slightly by appending the `ed25519Pubkey` field to the end of the struct so that parsers and the memory layout only need to change minimally.